kvstoremesh: correctly remove cached data upon cluster disconnection

giorio94 opened the pull request:
Currently, a comment attached to the Remove method claims that the cached data would have disappeared due to lease expiration. However, the lease is never expected to expire, as we use a single local client for all remote clusters. Additionally, depending on lease expiration would leave a significantly long race condition window in which stale data would still be present if the same cluster were to be reconnected.

In most cases, stale cached data is not a problem, as the etcd instance is stateless. However, it can create problems if the same cluster is first disconnected and then subsequently reconnected, without restarting the clustermesh-apiserver pod in between. Indeed, in this case we would have leftover entries that would not be cleaned up before reconnecting.

Let's address this issue by explicitly deleting the cached data upon cluster disconnection. This operation is performed synchronously (i.e., blocking other connection/disconnection operations) to prevent race conditions causing the deletion of incorrect data. Yet, to avoid blocking forever in case of errors, we only allow a limited number of retries, and give up with a loud error if they all fail.

<!-- Description of change -->

Fixes: #issue-number

```release-note
Correctly remove data cached by KVStoreMesh for a given cluster when disconnecting from such cluster
```

giorio94 committed:
kvstoremesh: correctly remove cached data upon cluster disconnection

Currently, a comment attached to the Remove method claims that the
cached data would have disappeared due to lease expiration. However,
the lease is never expected to expire, as we use a single local client
for all remote clusters. Additionally, depending on lease expiration
would leave a significantly long race condition window in which stale
data would still be present if the same cluster were to be reconnected.

In most cases, stale cached data is not a problem, as the etcd instance
is stateless. However, it can create problems if the same cluster is
first disconnected and then subsequently reconnected, without restarting
the clustermesh-apiserver pod in between. Indeed, in this case we would
have leftover entries that would not be cleaned up before reconnecting.

Let's address this issue by explicitly deleting the cached data upon
cluster disconnection. This operation is performed synchronously (i.e.,
blocking other connection/disconnection operations) to prevent race
conditions causing the deletion of incorrect data. Yet, to avoid blocking
forever in case of errors, we only allow a limited number of retries, and
give up with a loud error if they all fail.

As a safety measure, the removal of cached data can also be disabled
via the hidden "disable-drain-on-disconnection" flag.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

marseel commented on pkg/clustermesh/kvstoremesh/remote_cluster.go:
```diff
@@ -130,8 +133,74 @@ func (rc *remoteCluster) Stop() {
+				return err
+			}
+
+			prefixes = prefixes[1:]
```
nit: Is there any reason why we are overriding prefixes/keys?
I know it's technically correct, but a bit strange to read.

giorio94 replied:
The idea is to avoid attempting to delete again successful prefixes/keys in case an error has occurred at a later stage. That's only a minor optimization, I can definitely drop it if you prefer.

marseel commented on pkg/clustermesh/kvstoremesh/remote_cluster.go:
```diff
@@ -130,8 +133,74 @@ func (rc *remoteCluster) Stop() {
+		path.Join(kvstore.ClusterConfigPrefix, rc.name),
+	}
+	prefixes := []string{
+		path.Join(kvstore.SyncedPrefix, rc.name),
```
If I remember correctly, order in which we remove here is probably very important. 
I would love to see some reasoning of what happens to running cilium-agents and bootstrapping cilium-agents. 

For example, if we remove `ClusterConfigPrefix` first, then new clustermesh clients won't know that they should use syncCanaries, which could result in reading a partial state - which probably is fine, but it makes me wonder also about other cases.

giorio94 replied:
IMO deleting the cluster config first is the right thing to do here, because that effectively prevents bootstrapping agents from connecting (because they expect the config to be present), which is the expected behavior considering that we are deleting the cached data for that cluster in any case.

The deletion order of the other items doesn't matter once the agents are already connected, and they'll receive the deletion events for the various keys. Regardless, it is expected that agents will disconnect from the same cluster as well, hence they would drain all the keys in any case.

Another possibility would be that the bootstrapping agent already retrieved the cluster config, but hasn't started watching yet. In this case, deleting the synced prefix first prevents again from connecting, which looks correct to me, although that's arguably a rare condition.

marseel reviewed (changes requested):
one nit, one bigger comment. 
Anyway, if we figure out these corner cases, I would encourage to do it behind a hidden flag, just in case, we need to disable it due to some race-condition.

marseel reviewed (approved):
Thanks, looks great!

giorio94 committed:
clustermesh: make cluster disconnection asynchronous

Currently, both cluster configuration upsertions and removals are
handled synchronously. However, while upsertions are negligible as
they only trigger the restart of a controller, removals can take
a significant amount of time, possibly blocking any other parallel
operations. This applies both to the agents logic, due to the
draining of all previously known entries, and to kvstoremesh, as
the subsequent commits will implement the removal of the cached
data upon disconnection.

Let's introduce the support for processing the configurations of
separate clusters in parallel by running the removal logic
asynchronously. Still, we need to ensure that events for the same
cluster continue to be processed in order, to prevent issues in
case the configuration of a given cluster is first removed and
then subsequently added back again while the removal is still
in progress. In this case, we delay the addition event and replay
it only once the removal terminated.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

marseel commented on pkg/clustermesh/common/clustermesh.go:
```diff
@@ -108,13 +118,17 @@ func (cm *clusterMesh) Start(cell.HookContext) error {
 
+	// Wait until all in-progress removal processes have completed, if any.
+	// We must not hold the mutex at this point, as needed by the go routines.
+	cm.wg.Wait()
```
As you mention it below:
> // Run onRemove in a separate go routing as potentially slow, to avoid
> // blocking the processing of further events in the meanwhile.

We might need to wait quite some time on `cm.wg.Wait()`
I'm not sure this is a good idea, agent shutdown should probably not block on clustermesh's deleted cluster cleanup. 
I know it was like that before too, but probably we shouldn't wait on it during shutting down of agent at all?

Currently, it seems like we setting [terminationGracePeriodSeconds to 1](https://github.com/cilium/cilium/blob/c22eceb75bb4187b011bc665058adc4aba765b78/install/kubernetes/cilium/values.yaml#L142) :exploding_head: 
So while I think for clustermesh it doesn't matter that much if it will be cleaned or not, I wonder if for other components it might actually matter whether they execute or not as they execute sequentially:
https://github.com/cilium/cilium/blob/2311f3d7811c8950b7f5e1e2ac290f8216be0dc3/vendor/github.com/cilium/hive/cell/lifecycle.go#L137

giorio94 replied:
I think we need to distinguish different cases here:

* Plain agent or kvstoremesh shutdown, with no clustermesh configurations removals (i.e., the standard case): clustermesh just stops the watchers and closes the etcd client. This was, and continues to be, performed synchronously by the stop hook, and is expected to be instantaneous (unless the agent is blocked processing events already received).
* agents shutdown while the removal of one or more clusters is being processed. In this case, it is possible that the cleanup doesn't complete in time with such a short termination grace period, as the agents would need to finish processing all the synthetic deletion events. However, this was already the case earlier, and it is an extremely rare condition IMO.
* kvstoremesh shutdown while the removal of one or more clusters is being processed. In that case, [the context of the job should be closed](https://github.com/giorio94/cilium/blob/mio%2Fkvstoremesh-delete-cluster/pkg/clustermesh/kvstoremesh/remote_cluster.go#L183), hence causing the removal logic to return immediately, and unblock the wait group. Now that I'm writing this, I reckon there may be a startup/shutdown ordering issue that makes that not work as expected, let me double check and address it in case.

giorio94 replied:
> I reckon there may be a startup/shutdown ordering issue that makes that not work as expected, let me double check and address it in case.

Fixed with the latest push: https://github.com/cilium/cilium/compare/81721f9e913c55d75137a74c4bb0ec45d5815fc2..a5ad5df50a339422506088edab0b67c68336715a

marseel commented on pkg/clustermesh/common/clustermesh.go:
```diff
@@ -161,8 +175,21 @@ func (cm *clusterMesh) add(name, path string) {
+
+func (cm *clusterMesh) addLocked(name, path string) {
+	if _, ok := cm.tombstones[name]; ok {
+		// The configuration for this cluster has been recreated before that the cleanup
```
```suggestion
		// The configuration for this cluster has been recreated before the cleanup
```

marseel commented on pkg/clustermesh/common/clustermesh.go:
```diff
@@ -182,13 +208,44 @@ func (cm *clusterMesh) add(name, path string) {
+	cm.conf.Metrics.TotalRemoteClusters.WithLabelValues(cm.conf.ClusterInfo.Name, cm.conf.NodeName).Set(float64(len(cm.clusters)))
+
+	cm.wg.Add(1)
+	go func() {
```
I'm not sure if doing this async doesn't make it worse for the case when someone changes cluster-name, without changing ClusterID, no?
Previous cluster is going to be deleting while config with a new cluster name will fail on ClusterID reservation?

giorio94 replied:
I don't think there would be any difference, as the inotify watcher is non deterministic (as in, it may be, but it depends on the actual file names), and we could either first observe the addition or deletion event.

marseel replied:
makes sense, thanks for the explanation.

marseel reviewed (changes requested):
Overal makes sense, but I would question whether this part:
```
 To address the fact that the removal process now takes longer, the first commit makes this operation asynchronous, so that additions and removals of other cluster configurations can be processed in parallel. Differently, the recreation of a configuration for a cluster which is being removed is queued and replayed once the removal completes, to still ensure the correct ordering of the operations.
```
is necessary. Isn't that also increase the load on agent as we will be doing removal in parallel now? 

I'm honestly on the edge what would be better/safer. I don't think we should have super fast handling for cluster removal even if it blocks adding new clusters, if it introduces significant mental complexity around tombstones and removal of clusters.

This was mostly triggered due to the fact that I'm not really sure if using name vs clusterID in tombstones is a good idea.

marseel reviewed (approved):
I want to make sure we understand all the consequences of this change first - especially as this change touches both kvstoremesh and the agent side, before merging :) 

I would say that two examples that you gave:
```
agents shutdown while the removal of one or more clusters is being processed. In this case, it is possible that the cleanup doesn't complete in time with such a short termination grace period, as the agents would need to finish processing all the synthetic deletion events. However, this was already the case earlier, and it is an extremely rare condition IMO.
```
and
```
Blocking the configuration of a new cluster for a relatively long period of time would be confusing for users. Let's consider the kvstoremesh case with the 3 minutes grace period. Supposing a user disconnects 5 clusters and connects a new one, the new one may not be reconciled for 15 minutes, giving the impression that something didn't work. The order in which the cluster configuration events are processed is also undeterministic (it depends on the inotify), which makes it even worse.
```
have probably a similar frequency of happening. In general, adding/removing clusters does not happen often, but once it happens, churn on nodes is not something very unlikely to happen during this time, nodes come and go all the time.

I know it was like this before - blocking agent on cluster removal, so it's not something blocking ( :wink: )  this PR, but perhaps something to improve later on.

giorio94 committed:
clustermesh: allow aborting cluster removal via dedicated context

The removal of a cluster can take a significant amount of time, as in
the case of the kvstoremesh logic introduced in subsequent commits.
Hence, let's propagate a context that gets canceled upon clustermesh
shutdown, to allow immediately interrupting ongoing removal processes
if not yet terminated at that point.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

marseel reviewed (commented):
Nice, looks good. 

A bit confusing for me was that onRemove uses a different context from drainJobFn though and we don't cancel drainJobFn in Stop() of `func (cm *clusterMesh) Stop(cell.HookContext) error {`
If my understanding is correct these context are attached to two different Stop hooks, right? One from jobGroup thing, the other one to new rctx? I know it's still correct, but a bit confusing to me. 
Ideally, we would have rctx attached to jobGroup, no?

Just wanted to confirm my understanding, but lgtm otherwise.

marseel reviewed (approved):
very nice and easy to read now, double approve :100:

joestringer commented on pkg/clustermesh/kvstoremesh/kvstoremesh.go:
```diff
@@ -42,6 +45,9 @@ var DefaultConfig = Config{
 	flags.Duration("per-cluster-ready-timeout", def.PerClusterReadyTimeout, "Remote clusters will be disregarded for readiness checks if a connection cannot be established within this duration")
 	flags.Duration("global-ready-timeout", def.GlobalReadyTimeout, "KVStoreMesh will be considered ready even if any remote clusters have failed to synchronize within this duration")
+
+	flags.Bool("disable-drain-on-disconnection", def.DisableDrainOnDisconnection, "Do not drain cached data upon cluster disconnection")
```
Please prefer `enable-drain-on-disconnection` rather than `disable-drain-on-connection`. Defining options as negative is confusing in some languages due to the double negative (if you _don't disable_ foo, ie `--disable-foo=false`, does that mean it's enabled or just that it's not explicitly enabled?)

joestringer replied:
Alternative that may even be better: don't use `enable/disable` at all.

marseel replied:
I think we can still change the name of it - it is a hidden option as we did not expect anyone to use it, just as a fail-safe mechanism. 
But that's a really good point with naming, I was scratching my head a few times before with similar options too.

giorio94 replied:
> Please prefer enable-drain-on-disconnection rather than disable-drain-on-connection. Defining options as negative is confusing in some languages due to the double negative (if you don't disable foo, ie --disable-foo=false, does that mean it's enabled or just that it's not explicitly enabled?)

Hmm, I see your point. 

The reason for calling it `disable` is to avoid `true` by default boolean options, which also tend to be confusing and problematic in certain contexts (ie., a user may expect that entirely omitting `--enable-drain-on-disconnection=false` would be equivalent to specifying it, which would not be the case defaulting to true -- conversely --disable-drain-on-disconnection looks more intuitive to me). Although a bit unrelated, AFAIR a few true by default settings are also currently broken in the helm chart for similar reasons.

That said, as Marcel mentioned, this option is a temporary fail-safe (hidden) toggle not intended to be exposed to users, and I agree that we should still be able to change the name without particular consequences in case.

joestringer replied:
>  [...] a user may expect that entirely omitting --enable-drain-on-disconnection=false would be equivalent to specifying it, which would not be the case defaulting to true [...]

To a degree I get the concern around default-true boolean options, but that's also just inherently how features work - features are either enabled or disabled. Depending on the maturity of the feature and how useful it is to the average user, we may decide to enable or disable different flags by default. There's then two important aspects to the UX - can users easily understand the default posture of the setting? And can they override it themselves?

Here's where I'm coming from in context of this specific flag: If I specify `--disable-drain-on-disconnection=false`, that means: Please don't "Do not drain cached data upon cluster disconnection".

This point is not a big deal for hidden options, but I raise it to encourage discussion about how we can provide more consistent UX across different flags. My gut feeling is that unless there is a really strong argument in a specific scenario, using "disable" (or other negative modifiers) in a flag name is almost always more confusing than the reverse.
